File: app/src/main/java/com/donald/musictheoryapp/Question/TruthQuestion.java
package com.donald.musictheoryapp.Question;

import com.donald.musictheoryapp.Utils.JSONArrayUtil;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class TruthQuestion extends Question
{
    public static class Answer implements Question.Answer
    {
        public Boolean userAnswer;
        public boolean correctAnswer;

        public Answer(Boolean userAnswer, boolean correctAnswer)
        {
            this.userAnswer = userAnswer;
            this.correctAnswer = correctAnswer;
        }

        public JSONObject toJSON() throws JSONException
        {
            JSONObject object = new JSONObject();
            object.put("user_answer", userAnswer == null ? JSONObject.NULL : userAnswer);
            object.put("correct_answer", correctAnswer);
            return object;
        }

        public static Answer fromJSON(JSONObject object) throws JSONException
        {
            return new Answer(
                object.isNull("user_answer") ? null : object.getBoolean("user_answer"),
                object.getBoolean("correct_answer")
            );
        }

        @Override
        public boolean correct()
        {
            return userAnswer != null && userAnswer == correctAnswer;
        }
    }

    public static final String TRUE_ANSWER = "true";
    public static final String FALSE_ANSWER = "false";

    public Answer answer;

    public TruthQuestion() {}

    @Override
    public int points()
    {
        if(answer.correct()) return 1;
        return 0;
    }

    @Override
    public void acceptVisitor(QuestionVisitor visitor) { visitor.visit(this); }

    public static TruthQuestion fromJSON(JSONObject object, QuestionGroup group) throws JSONException
    {
        TruthQuestion question = new TruthQuestion();
        question.number = object.getInt("number");
        question.group = group;
        question.descriptions = JSONArrayUtil.descriptions(object);
        question.answer = Answer.fromJSON(object.getJSONObject("answer"));
        return question;
    }
}

File: app/src/main/java/com/donald/musictheoryapp/question/TruthQuestion.kt
package com.donald.musictheoryapp.question

import com.donald.musictheoryapp.util.getDescriptions
import com.donald.musictheoryapp.util.getInputHint
import org.json.JSONObject

class TruthQuestion(
    number: Int,
    descriptions: List<Description>,
    inputHint: String?,
    val answer: Answer
) : ChildQuestion(number, descriptions, inputHint) {

    override val points: Int
        get() = if (answer.correct) 1 else 0
    override val maxPoints = 1
    override val correctAnswerCount = 1

    override fun acceptVisitor(visitor: QuestionVisitor) {
        visitor.visit(this)
    }

    override fun copyNew(): TruthQuestion {
        return TruthQuestion(
            number,
            List(descriptions.size) { i -> descriptions[i].copy() },
            inputHint,
            answer.copyNew()
        )
    }

    override fun toPartialJson(): JSONObject {
        return JSONObject().apply {
            put(QUESTION_TYPE, Type.TRUTH.ordinal)
            put("answer", answer.toJson())
        }
    }

    companion object {
        const val TRUE_ANSWER = "true"
        const val FALSE_ANSWER = "false"
        fun fromJson(jsonObject: JSONObject): TruthQuestion {
            return TruthQuestion(
                number = jsonObject.getInt("number"),
                descriptions = jsonObject.getDescriptions(),
                inputHint = jsonObject.getInputHint(),
                answer = Answer.fromJson(jsonObject.getJSONObject("answer"))
            )
        }
    }

    class Answer(var userAnswer: Boolean?, val correctAnswer: Boolean) : ChildQuestion.Answer {

        override val correct: Boolean
            get() {
                return userAnswer == correctAnswer
            }

        fun toJson(): JSONObject {
            return JSONObject().apply {
                put("user_answer", if (userAnswer == null) JSONObject.NULL else userAnswer)
                put("correct_answer", correctAnswer)
            }
        }

        override fun copyNew(): Answer {
            return Answer(null, correctAnswer)
        }

        companion object {
            fun fromJson(jsonObject: JSONObject): Answer {
                return Answer(
                    if (jsonObject.isNull("user_answer")) null else jsonObject.getBoolean("user_answer"),
                    jsonObject.getBoolean("correct_answer")
                )
            }
        }

    }

}